pageserver/src/walreceiver.rs

```diff
@@ -201,7 +201,7 @@ fn walreceiver_main(
 
                     let new_checkpoint_bytes = encode_checkpoint(checkpoint);
                     if new_checkpoint_bytes != old_checkpoint_bytes {
-                        timeline.put_page_image(checkpoint_tag, Lsn(0), new_checkpoint_bytes);
+                        timeline.put_page_image(checkpoint_tag, Lsn(0), new_checkpoint_bytes)?;
```

Comment:
Why do we use Lsn(0) here?
I was thinking that we store checkpoint records for certain lsns, but as I got from the code, we always overwrite same record with lsn 0. How does it work if we want to branch (and get checkpoint data) at some LSN in the past?

It seems incorrect to me. I think we don't observe problems only because pg_resetwal covers them.
I'll try to come up with some test that illustrates this.

Reply:
Checkpoint  record is unversioned. It was done to minimize storage footprint for updating it. Otherwise we will have to store one more record per each commit. Which can be sensible.

My assumption was that checkpoint fields are monotonically updated. So if we want to initialize some old snapshot with new checkpoint record, we will only get some gap in OIDs/XIDs. But it seems to be not critical and should not cause any problems. I only not sure about oldestXid...

Reply:
If we don't advance oldestXid correctly, we'll fail to perform xid wraparound.
Now, pg_resetwal just resets it to FirstNormalTransactionId. 
It means that as soon as main branch will reach a xid wraparound 'half-turn', all newly created branches will face the 'prevent wraparound' situation because nextXid is too far from oldestXid.

Maybe we can update checkpoint record only when these 'oldest*' fields are updated?
I agree that leaping over monotonically increasing nextXid and nextOID looks harmless, even if strange.